crates/virtio-queue/src/mock.rs

```diff
@@ -371,16 +397,60 @@ impl<'a, M: GuestMemory> MockSplitQueue<'a, M> {
         q
     }
 
-    /// Writes a single descriptor chain to the memory object of the queue, at the beginning of the
+    /// Writes descriptors to the memory object of the queue, at the beginning of the
     /// descriptor table, and returns the associated `DescriptorChain` object.
-    pub fn build_desc_chain(&self, descs: &[Descriptor]) -> DescriptorChain<&M> {
-        self.add_desc_chain(descs, 0);
+    pub fn build_chain_of_descriptors(
+        &self,
+        descs: &[Descriptor],
+    ) -> Result<DescriptorChain<&M>, Error> {
+        for (idx, desc) in descs.iter().enumerate() {
+            let i = idx as u16;
 
-        self.create_queue(self.mem)
+            let desc = Descriptor::new(desc.addr().0, desc.len(), desc.flags(), desc.next());
+            if let Err(e) = self.desc_table().store(i, desc) {
+                return Err(e);
+            }
+        }
+
+        // Update the first available ring position.
+        let avail_idx: u16 = self
+            .mem
+            .read_obj(self.avail_addr().unchecked_add(2))
+            .unwrap();
+        self.mem
+            .write_obj(
+                0,
+                self.avail_addr().unchecked_add(
+                    VIRTQ_AVAIL_RING_HEADER_SIZE + avail_idx as u64 * VIRTQ_AVAIL_ELEMENT_SIZE,
+                ),
+            )
+            .unwrap();
+
+        // Increment `avail_idx`.
+        self.mem
+            .write_obj(avail_idx + 1, self.avail_addr().unchecked_add(2))
+            .unwrap();
+        Ok(self
+            .create_queue(self.mem)
             .iter()
             .unwrap()
             .next()
-            .expect("failed to build desc chain")
+            .expect("failed to build desc chain"))
+    }
+
+    /// Writes a single descriptor chain to the memory object of the queue, at the beginning of the
+    /// descriptor table, and returns the associated `DescriptorChain` object.
+    pub fn build_desc_chain(&self, descs: &[Descriptor]) -> Result<DescriptorChain<&M>, Error> {
+        if self.add_desc_chain(descs, 0).is_ok() {
+            Ok(self
+                .create_queue(self.mem)
+                .iter()
+                .unwrap()
```

Comment:
This function now returns a result, but we're still using `unwrap` and `expect`. Is there any reason for that?

I would also recommend trying to rewrite this using Rust combinators because it makes the code easier to read. Here are some examples:
- https://doc.rust-lang.org/rust-by-example/error/option_unwrap/question_mark.html
- https://doc.rust-lang.org/rust-by-example/error/option_unwrap/map.html
- https://doc.rust-lang.org/rust-by-example/error/option_unwrap/and_then.html

Reply:
I believe the unwrap should exist because the`iter` function returns a `Result<AvailIter<'_, M::T>, Error>` and the `next` function should be called on an `AvailIter` object. Also, the next function returns `Option<Self::Item>`, where `Item` is an alias for `DescriptorChain<M>`, which the function returns. Maybe this could be replaced with something else, but I don't know with what exactly.